init.lua

```diff
@@ -947,29 +892,60 @@ mt.register_on_joinplayer(function(player)
 	init_data(player, name)
 end)
 
-local function save_meta(player, data)
-	local meta = player:get_meta()
-	meta:set_string("inv_items", serialize(data.inv_items))
-end
-
 mt.register_on_leaveplayer(function(player)
 	local name = player:get_player_name()
-	if progressive_mode then
-		save_meta(player, player_data[name])
-	end
-
 	player_data[name] = nil
 end)
 
 if progressive_mode then
+	craftguide.add_recipe_filter("Progressive filter", function(recipes, player)
+		local name = player:get_player_name()
+		local discovered = player_data[name].inv_items
+		local inv_items = get_inv_items(player)
+		discovered = table_clean(table_merge(discovered, inv_items))
+
+		if #discovered == 0 then
+			return {}
+		end
+
+		local filtered = {}
+		for i = 1, #recipes do
+			local recipe = recipes[i]
+			local recipe_in_inv = true
+
+			for _, item in pairs(recipe.items) do
+				if not item_in_inv(item, discovered) then
+					recipe_in_inv = false
+				end
+			end
+
+			if recipe_in_inv then
+				filtered[#filtered + 1] = recipe
+			end
+		end
+
+		return filtered
+	end)
+
+	mt.register_on_joinplayer(function(player)
+		local meta = player:get_meta()
+		local inv_items = deserialize(meta:get_string("inv_items")) or {}
+		local name = player:get_player_name()
+		player_data[name].inv_items = inv_items
+	end)
+
+	local function save_meta(player)
+		local meta = player:get_meta()
+		local name = player:get_player_name()
+		meta:set_string("inv_items", serialize(player_data[name].inv_items))
+	end
+
+	mt.register_on_leaveplayer(save_meta)
```

Comment:
The inv items are saved whatsoever now, right?

Reply:
not sure what you mean? `save_meta` still saves the items.

Reply:
I'm referring to:

> Also previously saving of inventory items would happen even if the progressive filter was overridden

The behavior looks unchanged.

Reply:
Oh right. But now you can disable progressive mode and add your own filter instead.